src/pages/RFOX/hooks/useEarliestBlockNumberByTimestampQuery/useEarliestBlockNumberByTimestampQuery.ts

```diff
@@ -0,0 +1,109 @@
+import assert from 'assert'
+import { useMemo } from 'react'
+import { arbitrum } from 'viem/chains'
+import { useQuery } from 'wagmi/query'
+import { viemClientByNetworkId } from 'lib/viem-client'
+
+type BlockNumberByTimestampQueryKey = ['blockNumberByTimestamp', { targetTimestamp: bigint }]
+
+/**
+ * @param targetTimestamp The target timestamp in seconds - not an actual block timestamp
+ */
+type UseEarliestBlockNumberByTimestampQueryProps = {
+  targetTimestamp: bigint
+}
+
+const AVERAGE_BLOCK_TIME_BLOCK_DISTANCE = 10_000
+
+const client = viemClientByNetworkId[arbitrum.id]
+
+const calcNumberBlocksToMove = (timeDifferenceSeconds: bigint, averageBlockTimeSeconds: number) => {
+  const blocksToMove = BigInt(Math.ceil(Number(timeDifferenceSeconds) / averageBlockTimeSeconds))
+
+  return blocksToMove
+}
+
+export const getEarliestBlockNumberByTimestampQueryKey = ({
+  targetTimestamp,
+}: UseEarliestBlockNumberByTimestampQueryProps): BlockNumberByTimestampQueryKey => [
+  'blockNumberByTimestamp',
+  { targetTimestamp },
+]
+
+export const getEarliestBlockNumberByTimestampQueryFn =
+  ({
+    targetTimestamp,
+    averageBlockTimeBlockDistance = AVERAGE_BLOCK_TIME_BLOCK_DISTANCE,
+  }: UseEarliestBlockNumberByTimestampQueryProps & { averageBlockTimeBlockDistance?: number }) =>
+  async () => {
+    const latestBlock = await client.getBlock()
+
+    assert(targetTimestamp <= latestBlock.timestamp, 'Target timestamp must not be in the future')
+
+    const historicalBlock = await client.getBlock({
+      blockNumber: latestBlock.number - BigInt(averageBlockTimeBlockDistance),
+    })
+
+    const averageBlockTimeSeconds =
+      Number(latestBlock.timestamp - historicalBlock.timestamp) / averageBlockTimeBlockDistance
+    const durationSeconds = latestBlock.timestamp - targetTimestamp
+    const targetBlocksToMove = calcNumberBlocksToMove(durationSeconds, averageBlockTimeSeconds)
+
+    let blockNumber = latestBlock.number - targetBlocksToMove
+
+    while (true) {
+      if (blockNumber <= 0n) {
+        return 0n
+      }
+
+      const block = await client.getBlock({ blockNumber })
+      const timeDifferenceSeconds = targetTimestamp - block.timestamp
+
+      // Block is within 1 block before the target timestamp
+      if (timeDifferenceSeconds >= 0n && timeDifferenceSeconds <= averageBlockTimeSeconds) {
+        break
+      }
+
+      const blocksToMove = calcNumberBlocksToMove(timeDifferenceSeconds, averageBlockTimeSeconds)
```

Comment:
noting that we lost the absolute value for `timeDifferenceSeconds` here. feels like it will still result in finding the correct block, but may take an extra iteration due to using math.Ceil in the blocks to move calculation. (ex. -20.5 -> -20, vs 20.5 -> 21). just noting, the slight difference in behavior

Reply:
Nice, I've actioned this by re-incorporating `Math.abs` but with `floor` instead of `ceil` due to an infinite loop detected in the unit tests - due to the sub-second block times it was jumping either side of the target block timestamp infinitely so I used floor instead and we're getting every target block with 1-2 iterations max.

7dc78e0fd5165d677fad538d25bee275eed8a340